From f132afedb99a2fa691063e151df6b95f2675f645 Mon Sep 17 00:00:00 2001
From: Yongjun Zhang <yzhang@cloudera.com>
Date: Fri, 8 May 2015 00:11:46 -0700
Subject: [PATCH 1114/2848] HDFS-8311. DataStreamer.transfer() should timeout
 the socket InputStream. (Esteban Gutierrez via
 Yongjun Zhang)

(cherry picked from commit 730f9930a48259f34e48404aee51e8d641cc3d36)
(cherry picked from commit fbb2cf54bcaa4ce739747e04fce4058b2639a619)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java

Change-Id: I0d8e96aa5655b9a504a08e92fc33a5b9aff27e22
---
 .../org/apache/hadoop/hdfs/DFSOutputStream.java    |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
index 11eb5f7..85264fc 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSOutputStream.java
@@ -1244,9 +1244,10 @@ private void transfer(final DatanodeInfo src, final DatanodeInfo[] targets,
       try {
         sock = createSocketForPipeline(src, 2, dfsClient);
         final long writeTimeout = dfsClient.getDatanodeWriteTimeout(2);
+        final long readTimeout = dfsClient.getDatanodeReadTimeout(2);
         
         OutputStream unbufOut = NetUtils.getOutputStream(sock, writeTimeout);
-        InputStream unbufIn = NetUtils.getInputStream(sock);
+        InputStream unbufIn = NetUtils.getInputStream(sock, readTimeout);
         IOStreamPair saslStreams = dfsClient.saslClient.socketSend(sock,
           unbufOut, unbufIn, dfsClient, blockToken, src);
         unbufOut = saslStreams.out;
@@ -1529,9 +1530,10 @@ private boolean createBlockOutputStream(DatanodeInfo[] nodes,
           assert null == blockReplyStream : "Previous blockReplyStream unclosed";
           s = createSocketForPipeline(nodes[0], nodes.length, dfsClient);
           long writeTimeout = dfsClient.getDatanodeWriteTimeout(nodes.length);
+          long readTimeout = dfsClient.getDatanodeReadTimeout(nodes.length);
           
           OutputStream unbufOut = NetUtils.getOutputStream(s, writeTimeout);
-          InputStream unbufIn = NetUtils.getInputStream(s);
+          InputStream unbufIn = NetUtils.getInputStream(s, readTimeout);
           IOStreamPair saslStreams = dfsClient.saslClient.socketSend(s,
             unbufOut, unbufIn, dfsClient, accessToken, nodes[0]);
           unbufOut = saslStreams.out;
-- 
1.7.9.5

